### PR TITLE
GUAC-1101: Rewrite guacamole-auth-mysql with speed in mind.

### PR DESCRIPTION
**Pull request by mike-jumper**
:elephant: 2 of 2

This change is a massive rewrite of guacamole-auth-mysql. It adds support for the new batching guacamole-ext operations, and has been reorganized and cleaned up such that:

1. The threading issues are gone.
2. It can be easily modified to fit any database backend.

The authentication provider now exists as a larger parent project, guacamole-auth-jdbc, which contains a base project and a MySQL implementation. The base project contains 99% of the code, while the database-specific implementation contains only the database-specific configuration and SQL mapping XML. Adding support for a new database now really only requires porting a few SQL queries.

Permissions logic has been moved to the lowest level possible, and almost all shared functionality between objects now resides in abstract superclasses and generics.

### CODE REVIEW

File: extensions/guacamole-auth-jdbc/modules/guacamole-auth-jdbc-base/src/main/java/org/glyptodon/guacamole/auth/jdbc/connectiongroup/ConnectionGroupService.java [hunk 1]
```diff
@@ -0,0 +1,227 @@
...
+
+    @Override
+    protected void validateNewObject(AuthenticatedUser user, ConnectionGroup object)
+            throws GuacamoleException {
+
+        // Name must not be blank
+        if (object.getName().trim().isEmpty())
+            throw new GuacamoleClientException("Connection group names must not be blank.");
+        
+        // FIXME: Do not attempt to create duplicate connection groups
```
Comment (jmuehlner) [inline extensions/guacamole-auth-jdbc/modules/guacamole-auth-jdbc-base/src/main/java/org/glyptodon/guacamole/auth/jdbc/connectiongroup/ConnectionGroupService.java]: What's stopping us from implementing this now?

File: extensions/guacamole-auth-jdbc/modules/guacamole-auth-jdbc-base/src/main/java/org/glyptodon/guacamole/auth/jdbc/connectiongroup/ConnectionGroupService.java [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+
+    @Override
+    protected void validateExistingObject(AuthenticatedUser user,
+            ModeledConnectionGroup object) throws GuacamoleException {
+
+        // Name must not be blank
+        if (object.getName().trim().isEmpty())
+            throw new GuacamoleClientException("Connection group names must not be blank.");
+        
+        // FIXME: Check whether such a connection group is already present
```
Comment (jmuehlner) [inline extensions/guacamole-auth-jdbc/modules/guacamole-auth-jdbc-base/src/main/java/org/glyptodon/guacamole/auth/jdbc/connectiongroup/ConnectionGroupService.java]: ?

File: extensions/guacamole-auth-jdbc/modules/guacamole-auth-jdbc-base/src/main/java/org/glyptodon/guacamole/auth/jdbc/connection/ConnectionService.java [hunk 3]
```diff
@@ -0,0 +1,407 @@
...
+
+    @Override
+    protected void validateNewObject(AuthenticatedUser user, Connection object)
+            throws GuacamoleException {
+
+        // Name must not be blank
+        if (object.getName().trim().isEmpty())
+            throw new GuacamoleClientException("Connection names must not be blank.");
+        
+        // FIXME: Do not attempt to create duplicate connections
```
Comment (jmuehlner) [inline extensions/guacamole-auth-jdbc/modules/guacamole-auth-jdbc-base/src/main/java/org/glyptodon/guacamole/auth/jdbc/connection/ConnectionService.java]: ?

File: extensions/guacamole-auth-jdbc/modules/guacamole-auth-jdbc-base/src/main/java/org/glyptodon/guacamole/auth/jdbc/connection/ConnectionService.java [hunk 4]
```diff
@@ -0,0 +1,407 @@
...
+
+    @Override
+    protected void validateExistingObject(AuthenticatedUser user,
+            ModeledConnection object) throws GuacamoleException {
+
+        // Name must not be blank
+        if (object.getName().trim().isEmpty())
+            throw new GuacamoleClientException("Connection names must not be blank.");
+        
+        // FIXME: Check whether such a connection is already present
```
Comment (jmuehlner) [inline extensions/guacamole-auth-jdbc/modules/guacamole-auth-jdbc-base/src/main/java/org/glyptodon/guacamole/auth/jdbc/connection/ConnectionService.java]: ?

File: extensions/guacamole-auth-jdbc/modules/guacamole-auth-jdbc-base/src/main/java/org/glyptodon/guacamole/auth/jdbc/connectiongroup/ConnectionGroupService.java (hunk 1 above)
Comment (mike-jumper) [inline extensions/guacamole-auth-jdbc/modules/guacamole-auth-jdbc-base/src/main/java/org/glyptodon/guacamole/auth/jdbc/connectiongroup/ConnectionGroupService.java]: Nothing but the desire to keep the review from getting huge beyond belief (not that it isn't already).

**Pull request merged**